geomesa-z3/src/main/scala/org/locationtech/geomesa/curve/XZ3SFC.scala

```diff
@@ -0,0 +1,447 @@
+        val (min, max) = sequenceInterval(oct.xmin, oct.ymin, oct.zmin, level, partial = false)
+        ranges.add(IndexRange(min, max, contained = true))
+      } else if (isOverlapped(oct)) {
+        // some portion of this range is excluded
+        // add the partial match and queue up each sub-range for processing
+        val (min, max) = sequenceInterval(oct.xmin, oct.ymin, oct.zmin, level, partial = true)
+        ranges.add(IndexRange(min, max, contained = false))
+        oct.children.foreach(remaining.add)
+      }
+    }
+
+    // initial level
+    LevelOneElements.foreach(remaining.add)
+    remaining.add(LevelTerminator)
+
+    // level of recursion
+    var level: Short = 1
+
+    while (level < g && !remaining.isEmpty && ranges.size < rangeStop) {
+      val next = remaining.poll
+      if (next.eq(LevelTerminator)) {
+        // we've fully processed a level, increment our state
+        if (!remaining.isEmpty) {
+          level = (level + 1).toShort
+          remaining.add(LevelTerminator)
+        }
+      } else {
+        checkValue(next, level)
+      }
+    }
+
+    // bottom out and get all the ranges that partially overlapped but we didn't fully process
+    while (!remaining.isEmpty) {
+      val oct = remaining.poll
+      if (oct.eq(LevelTerminator)) {
+        level = (level + 1).toShort
+      } else {
+        val (min, max) = sequenceInterval(oct.xmin, oct.ymin, oct.zmin, level, partial = false)
+        ranges.add(IndexRange(min, max, contained = false))
+      }
+    }
+
+    // we've got all our ranges - now reduce them down by merging overlapping values
+    // note: we don't bother reducing the ranges as in the XZ paper, as accumulo handles lots of ranges fairly well
+    ranges.sort(IndexRange.IndexRangeIsOrdered)
+
+    var current = ranges.get(0) // note: should always be at least one range
+    val result = ArrayBuffer.empty[IndexRange]
+    var i = 1
+    while (i < ranges.size()) {
+      val range = ranges.get(i)
+      if (range.lower <= current.upper + 1) {
+        // merge the two ranges
+        current = IndexRange(current.lower, math.max(current.upper, range.upper), current.contained && range.contained)
+      } else {
+        // append the last range and set the current range for future merging
+        result.append(current)
+        current = range
+      }
+      i += 1
+    }
+    // append the last range - there will always be one left that wasn't added
+    result.append(current)
+
+    result
+  }
+
+  /**
+    * Computes the sequence code for a given point - for polygons this is the lower-left corner.
+    *
+    * Based on Definition 2 from the XZ-Ordering paper
+    *
+    * @param x normalized x value [0,1]
+    * @param y normalized y value [0,1]
+    * @param z normalized z value [0,1]
+    * @param length length of the sequence code that will be generated
+    * @return
+    */
+  private def sequenceCode(x: Double, y: Double, z: Double, length: Int): Long = {
```

Comment:
Should there maybe unit tests for this? and other functions in this class like `index`. I assume the math holds up here, but it's hard to be sure just by looking at it. There are unit tests which test range generation, but I'm not sure that covers this.